File: my-folder/0304-range-sum-query-2d---immutable/solution.ts
class NumMatrix {
    private prefix : number[][];

    constructor(matrix: number[][]) {
      const n : number = matrix.length, m : number = matrix[0].length;
      this.prefix = Array.from({ length : n + 1}, () => new Array(m + 1).fill(0));

      for(let i = 1; i <= n; i++){
          for(let j = 1; j <= m; j++){
            this.prefix[i][j] = this.prefix[i - 1][j] + this.prefix[i][j - 1] - this.prefix[i - 1][j - 1] + matrix[i - 1][j - 1];
          }
      }
        
    }

    sumRegion(row1: number, col1: number, row2: number, col2: number): number {
         return (
            this.prefix[row2 + 1][col2 + 1] -
            this.prefix[row1][col2 + 1] -
            this.prefix[row2 + 1][col1] +
            this.prefix[row1][col1]
        );
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * var obj = new NumMatrix(matrix)
 * var param_1 = obj.sumRegion(row1,col1,row2,col2)
 */

File: my-folder/0304-range-sum-query-2d---immutable/solution.java
class NumMatrix {
    private int[][] dp;

    public NumMatrix(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) return;
        
        int m = matrix.length;
        int n = matrix[0].length;
        // Allocate 1 extra size for easy boundary handling
        dp = new int[m + 1][n + 1]; 
        
        // Build the prefix sum matrix
        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                dp[r + 1][c + 1] = matrix[r][c] 
                                 + dp[r][c + 1] 
                                 + dp[r + 1][c] 
                                 - dp[r][c];
            }
        }
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        // Apply inclusion-exclusion principle using 1-indexed DP coordinates
        return dp[row2 + 1][col2 + 1] 
             - dp[row1][col2 + 1] 
             - dp[row2 + 1][col1] 
             + dp[row1][col1];
    }
}


/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */
